rust/core/src/driver_manager.rs

```diff
@@ -0,0 +1,1218 @@
+}
+
+impl Driver for ManagedDriver {
+    type DatabaseType = ManagedDatabase;
+
+    fn new_database(&mut self) -> Result<Self::DatabaseType> {
+        self.new_database_with_opts([])
+    }
+
+    fn new_database_with_opts(
+        &mut self,
+        opts: impl IntoIterator<Item = (<Self::DatabaseType as Optionable>::Option, OptionValue)>,
+    ) -> Result<Self::DatabaseType> {
+        let driver = &self.inner.driver.lock().unwrap();
+        let mut database = ffi::FFI_AdbcDatabase::default();
+
+        // DatabaseNew
+        let mut error = ffi::FFI_AdbcError::with_driver(driver);
+        let method = driver_method!(driver, DatabaseNew);
+        let status = unsafe { method(&mut database, &mut error) };
+        check_status(status, error)?;
+
+        // DatabaseSetOption
+        for (key, value) in opts {
+            set_option_database(driver, &mut database, self.inner.version, key, value)?;
+        }
+
+        // DatabaseInit
+        let mut error = ffi::FFI_AdbcError::with_driver(driver);
+        let method = driver_method!(driver, DatabaseInit);
+        let status = unsafe { method(&mut database, &mut error) };
+        check_status(status, error)?;
+
+        let inner = Arc::new(ManagedDatabaseInner {
+            database: Mutex::new(database),
+            driver: self.inner.clone(),
+        });
+        Ok(Self::DatabaseType { inner })
+    }
+}
+
+fn set_option_database(
+    driver: &ffi::FFI_AdbcDriver,
+    database: &mut ffi::FFI_AdbcDatabase,
+    version: AdbcVersion,
+    key: impl AsRef<str>,
+    value: OptionValue,
+) -> Result<()> {
+    let key = CString::new(key.as_ref())?;
+    let mut error = ffi::FFI_AdbcError::with_driver(driver);
+    let status = match (version, value) {
+        (_, OptionValue::String(value)) => {
+            let value = CString::new(value)?;
+            let method = driver_method!(driver, DatabaseSetOption);
+            unsafe { method(database, key.as_ptr(), value.as_ptr(), &mut error) }
+        }
+        (AdbcVersion::V110, OptionValue::Bytes(value)) => {
+            let method = driver_method!(driver, DatabaseSetOptionBytes);
+            unsafe {
+                method(
+                    database,
+                    key.as_ptr(),
+                    value.as_ptr(),
+                    value.len(),
+                    &mut error,
+                )
+            }
+        }
+        (AdbcVersion::V110, OptionValue::Int(value)) => {
+            let method = driver_method!(driver, DatabaseSetOptionInt);
+            unsafe { method(database, key.as_ptr(), value, &mut error) }
+        }
+        (AdbcVersion::V110, OptionValue::Double(value)) => {
+            let method = driver_method!(driver, DatabaseSetOptionDouble);
+            unsafe { method(database, key.as_ptr(), value, &mut error) }
+        }
+        (AdbcVersion::V100, _) => Err(Error::with_message_and_status(
+            ERR_ONLY_STRING_OPT,
+            Status::NotImplemented,
+        ))?,
```

Comment:
Maybe move this up to be an early return if `version == AdbcVersion::V100 && matches!(value, OptionValue::String(_))`? Makes it a bit easier to understand maybe?

Reply:
To be honest, I don't find this to be easier to understand. I prefer to have an exhaustive match without any `unreachable!` case. But I'm not strongly attached to it...